pipelines: updated KFP component concept page (#4062)
kubeflow/website: content/en/docs/components/pipelines/concepts/component.md

title = "Component"
description = "Conceptual overview of components in Kubeflow Pipelines"
weight = 20
+++

A **pipeline component** is the fundamental building block for an ML engineer to construct a Kubeflow Pipelines [pipeline][pipeline]. The component structure serves the purpose of packaging a functional unit of code along with its dependencies, so that it can be run as part of a workflow in a Kubernetes environement. Components can be combined in a [pipeline][pipeline] that creates a repeatable workflow, with individual components coordinating on inputs and outputs like parameters and [artifacts][artifacts].

A component is similar to a programming function. Indeed, it is most often implemented as a wrapper to a Python function using the [KFP Python SDK][KFP SDK]. However, a KFP component goes further than a simple function, with support for code dependencies, runtime environments, and distributed execution requirements.

> KFP components are designed to simplify constructing and running ML workflows in a Kubernetes environment. Using KFP components engineers can iterate faster, reduce maintenance overhead, and focus more attention on ML work.

Reviewer: ```suggestion
KFP components are designed to simplify constructing and running ML workflows in a Kubernetes environment. Using KFP components, engineers can iterate faster, reduce maintenance overhead, and focus more attention on ML work.
```

- Managing **code dependencies** (Python libraries and versions)
- Handling **system dependencies** (OS-level packages, GPU drivers, runtime environments)
- **Building and maintaining container images** and everything around this from container registry support to CVE (Common Vulnerabilities and Exposures) fixes
- **Deploying supporting resources** like PersistentVolumeClaims and ConfigMaps
- Handling **inputs and outputs** including metadata, parameters, logs, and artifacts
- **Ensuring compatibility** across clusters, images, and dependencies

KFP components simplify these challenges by enabling ML engineers to:
- **Stay at the Python level** - where most modern ML work occurs
- **Iterate quickly** – modify code without creating or rebuilding containers at each step
- **Focus on ML tasks** - rather than on platform and infrastructure concerns
- **Work seamlessly with Python IDE tools** – enable debugging, syntax highlighting, type checking, and docstring usage
- **Move between environments** – transition from local development to distributed execution with minimal changes

## What Does a Component Consist Of?

A KFP component consist of the following key elements:

### 1. Code

> - Typically a Python function, but can be other code such as a Bash command.

Reviewer: It'd be helpful to make it clear that you can run any binary if you use the container component type.

Author: I added a comment on this in the section on YAML components.

### 2. Dependency Support
- **Python libraries** - to be installed at runtime
- **Environment variables** - to be available in the runtime environment
- **Python package indices** - (for example, private PyPi servers) if needed to support installations
- **Cluster resources** - to support use of ConfigMaps, Secrets, PersistentVolumeClaims, and more
- **Runtime dependencies** - to support CPU, memory, and GPU requests and limits

### 3. Base Image

> - Defines the base container runtime environment

Reviewer: It'd be good to mention that the default base image is a generic Python environment.

Author: done

- May include system dependencies and pre-installed Python libraries

### 4. Input/Output (I/O) Specification
- Individual components cannot share in-memory data with each other, so they use the following concepts to support exchanging information and publishing results:
  - **Parameters** – for small values
  - **[Artifacts][artifacts]** - for larger data like model files, processed datasets, and metadata

## Constructing a Component

> ### 1. Python-Based Components

Reviewer 2: I wonder if it'd make sense to allow for imports to be global and applied to all components.

> ### 1. Python-Based Components

Reviewer: I think the downside is that perhaps not all components have a particular dependency available such as components using different base images or different components having different `packages_to_install` values. There is probably room to do improvements around this though.

> ### 1. Python-Based Components

Author: This is a good thought @mprahl to include -- imports confuse a lot of people as they are a limitation vs regular python. 

@franciscojavierarceo the idea of a `packages_to_install` parameter in the @pipeline decorator that would propagate to all components is intriguing. Could be useful for other component params too (like `pip_index_urls` and even `base image`)

> ### 1. Python-Based Components

Author: I added some clarification on this point in the section on Python-based components. I also linked to some of the user guides that mention this.

```python
from kfp.dsl import component, Output, Dataset

# hello component
@component()
def hello_world(name: str = "World") -> str:
    print(f"Hello {name}!")
    return name

# process data component
@component(
    base_image="python:3.12-slim-bookworm",
    packages_to_install=["pandas>=2.2.3"],
)
def process_data(output_data: Output[Dataset]):
    '''Get max from an array'''
    import pandas as pd
   # create dataset to write to output
    data = pd.DataFrame(data=[[1,2,3],[4,5,6]], columns=["a","b","c"])
    data.to_csv(output_data.path)
```

Observe that these are wrapped Python functions. The `@component` wrapper helps the KFP Python SDK supply the needed context for running these functions in containers as part of a KFP [pipeline][pipeline].

The `hello_world` component just uses default behavior (run the Python function on the default base image, which is `python:3.9`). The `process_data` component adds layers of customization, by supplying the name of a specific `base_image`, and `packages_to_install`. This component also uses KFP's `Output[Dataset]` class, which takes care of creating a KFP [artifact][artifact] type output.

> Note that inputs and outputs are defined as Python function parameters. Also, dependencies can often be installed at runtime, avoiding the need for custom base containers. Python-based components give close access to the Python tools that ML experimenters rely on, like modules and imports, usage information, type hints, and debugging tools.

Reviewer: ```suggestion
Note that inputs and outputs are defined as Python function parameters. Also, dependencies can often be installed at runtime, avoiding the need for custom base images. Python-based components give close access to the Python tools that ML experimenters rely on, like modules and imports, usage information, type hints, and debugging tools.
```

title = "Component"
description = "Conceptual overview of components in Kubeflow Pipelines"
weight = 20
+++

> A **pipeline component** is the fundamental building block for ML engineers to construct a Kubeflow Pipelines [pipeline][pipeline]. The component structure serves the purpose of packaging a functional unit of code along with its dependencies, so that it can be run as part of a workflow in a Kubernetes-based environement. Components can be combined in a [pipeline][pipeline] that creates a repeatable workflow, with individual components coordinating on inputs and outputs like parameters and [artifacts][artifacts].

Reviewer 2: ```suggestion
A **pipeline component** is the fundamental building block for an ML engineer to construct a Kubeflow [pipeline][pipeline]. The component structure serves the purpose of packaging a functional unit of code along with its dependencies, so that it can be run as part of a workflow in a Kubernetes environment. Components can be combined in a [pipeline][pipeline] that creates a repeatable workflow, with individual components coordinating on inputs and outputs like parameters and [artifacts][artifacts].
```

Author: Thanks for the suggestions @franciscojavierarceo. I used "Kubernetes-based", because pipelines can also be run on the Vertex AI Pipelines backend. Just "Kubernetes" is simpler, and I can see the justification. If any Googlers out there have a strong opinion on this, please let us know.

Author: Updated.

title = "Component"
description = "Conceptual overview of components in Kubeflow Pipelines"
weight = 20
+++

A **pipeline component** is the fundamental building block for an ML engineer to construct a Kubeflow Pipelines [pipeline][pipeline]. The component structure serves the purpose of packaging a functional unit of code along with its dependencies, so that it can be run as part of a workflow in a Kubernetes environement. Components can be combined in a [pipeline][pipeline] that creates a repeatable workflow, with individual components coordinating on inputs and outputs like parameters and [artifacts][artifacts].

A component is similar to a programming function. Indeed, it is most often implemented as a wrapper to a Python function using the [KFP Python SDK][KFP SDK]. However, a KFP component goes further than a simple function, with support for code dependencies, runtime environments, and distributed execution requirements.

> KFP components are designed to simplify constructing and running ML workflows in a Kubernetes environment. Using KFP components engineers can iterate faster, reduce maintenance overhead, and focus more attention on ML work.

Reviewer 3: Nit: after `Using KFP components` there should be a comma.

- **Environment variables** - to be available in the runtime environment
- **Python package indices** - (for example, private PyPi servers) if needed to support installations
- **Cluster resources** - to support use of ConfigMaps, Secrets, PersistentVolumeClaims, and more
- **Runtime dependencies** - to support CPU, memory, and GPU requests and limits

### 3. Base Image
- Defines the base container runtime environment (defaults to a generic Python base image)
- May include system dependencies and pre-installed Python libraries

### 4. Input/Output (I/O) Specification
- Individual components cannot share in-memory data with each other, so they use the following concepts to support exchanging information and publishing results:
  - **Parameters** – for small values
  - **[Artifacts][artifacts]** - for larger data like model files, processed datasets, and metadata

## Constructing a Component

### 1. Python-Based Components
The recommended way to define a component is using the `@dsl.component` decorator from the [KFP Python SDK][KFP SDK]. Below are two basic component definition examples:
```python
from kfp.dsl import component, Output, Dataset

> # hello component

Reviewer 3: Slight nit: For the first example, I think the comment should be `hello_world component` to map to the function name and match the second example.

Author: Great suggestions. Added.

@component()
def hello_world(name: str = "World") -> str:
    print(f"Hello {name}!")
    return name

# process data component
@component(
    base_image="python:3.12-slim-bookworm",
    packages_to_install=["pandas>=2.2.3"],
)
def process_data(output_data: Output[Dataset]):
    '''Get max from an array'''
    import pandas as pd
   # create dataset to write to output
    data = pd.DataFrame(data=[[1,2,3],[4,5,6]], columns=["a","b","c"])
    data.to_csv(output_data.path)
```

Observe that these are wrapped Python functions. The `@component` wrapper helps the KFP Python SDK supply the needed context for running these functions in containers as part of a KFP [pipeline][pipeline].

> The `hello_world` component just uses the default behavior, which is to run the Python function on the default base image (`python:3.9`). 

Reviewer: This might be hard to keep up to date. Perhaps mentioning that there is a default image is sufficient.

Author: Good call @mprahl . I replaced the static image with the command to get the default--which should be more stable.

> The `process_data` component adds layers of customization, by supplying the name of a specific `base_image`, and `packages_to_install`. Note the inclusion of the `import pandas as pd` statement inside the function; since the function will run inside a container (and won't have the script context), all non-builtin Python library dependencies need to be imported within the component function. This component also uses KFP's `Output[Dataset]` class, which takes care of creating a KFP [artifact][artifact] type output. 

Reviewer 3: When I view this page [here](https://github.com/dandawg/website/blob/update-concept-components/content/en/docs/components/pipelines/concepts/component.md) the "artifact" link is broken.

Author: Ah. I had it plural on the others. Will fix.

> The `process_data` component adds layers of customization, by supplying the name of a specific `base_image`, and `packages_to_install`. Note the inclusion of the `import pandas as pd` statement inside the function; since the function will run inside a container (and won't have the script context), all non-builtin Python library dependencies need to be imported within the component function. This component also uses KFP's `Output[Dataset]` class, which takes care of creating a KFP [artifact][artifacts] type output. 

Reviewer: It was confusing me since you still need to import the Python standard library.
```suggestion
The `process_data` component adds layers of customization, by supplying the name of a specific `base_image`, and `packages_to_install`. Note the inclusion of the `import pandas as pd` statement inside the function; since the function will run inside a container (and won't have the script context), all Python library dependencies need to be imported within the component function. This component also uses KFP's `Output[Dataset]` class, which takes care of creating a KFP [artifact][artifacts] type output. 
```

### 2. YAML-Based Components

The KFP backend uses YAML-based definitions to specify components. While the [KFP Python SDK][KFP SDK] can do this conversion automatically when a Python-based [pipeline][pipeline] is submitted, some use-cases can benefit from the direct YAML-based component approach.

A YAML-based component definition has the following parts:

* **Metadata:** name, description, etc.
* **Interface:** input/output specifications (name, type, description, default value, etc).
* **Implementation:** A specification of how to run the component given a set of argument values for the component’s inputs. The implementation section also describes how to get the output values from the component once the component has finished running.

YAML-based components support system commands directly. In fact, any command (or binary) that exists on the base image can be run. Here is simple YAML-based component example:
```yaml
# my_component.yaml file
name: my-component

> description:

Reviewer 3: Looks like a description value is missing here

Author: Good find. I meant to update this and never did. I gave it a simple description.

inputs:
- {name: string prefix, type: String}
- {name: num, type: Integer}

> outputs:

Reviewer 3: Should the value here be an empty list `[]` ?
